hathor/indexes/manager.py

```diff
@@ -66,23 +110,82 @@ def enable_tokens_index(self) -> None:
 
-            self.all_tips.add_tx(tx)
+        db_last_started_at = tx_storage.get_last_started_at()
+        sorted_all_db_name = self.sorted_all.get_db_name()
+        if sorted_all_db_name is None:
+            can_use_timestamp_index = False
+        else:
+            sorted_all_index_last_started_at = tx_storage.get_index_last_started_at(sorted_all_db_name)
+            can_use_timestamp_index = db_last_started_at == sorted_all_index_last_started_at
+
+        indexes_to_init: List[Tuple[_IndexFilter, BaseIndex]] = []
+        for index_filter, index in self._iter_all_indexes_with_filter():
+            index_db_name = index.get_db_name()
+            if index_db_name is None:
+                indexes_to_init.append((index_filter, index))
+                continue
+            index_last_started_at = tx_storage.get_index_last_started_at(index_db_name)
+            if db_last_started_at != index_last_started_at:
+                indexes_to_init.append((index_filter, index))
 
+        if indexes_to_init:
+            self.log.debug('there are indexes that need initialization',
+                           indexes_to_init=[i for _, i in indexes_to_init])
+        else:
+            self.log.debug('there are no indexes that need initialization')
+
+        iter_tx: Iterator[BaseTransaction]
+        if can_use_timestamp_index:
+            iter_tx = tx_storage._topological_sort_timestamp_index()
+        # XXX: this mostly exist because some tests will break when not using a caching-storage because of metadata
+        #      objects being duplicated which make the topological metadata iterator not work
+        elif tx_storage.is_caching_storage():
+            iter_tx = tx_storage._topological_sort_metadata()
+        else:
+            iter_tx = tx_storage._topological_sort_dfs()
+
+        # make sure that all the indexes that we're rebuilding are cleared
+        for _, index in indexes_to_init:
+            index_db_name = index.get_db_name()
+            if index_db_name:
+                tx_storage.set_index_last_started_at(index_db_name, 0)
+            index.force_clear()
+
+        block_count = 0
+        tx_count = 0
+        latest_timestamp = BLOCK_GENESIS.timestamp
+        first_timestamp = BLOCK_GENESIS.timestamp
+
+        for tx in progress(iter_tx, log=self.log):
+            # XXX: these would probably make more sense to be their own simple "indexes" instead of how it is here
+            latest_timestamp = max(tx.timestamp, latest_timestamp)
+            first_timestamp = min(tx.timestamp, first_timestamp)
             if tx.is_block:
-                self.block_tips.add_tx(tx)
-            elif not tx_meta.voided_by:
-                self.tx_tips.add_tx(tx)
+                block_count += 1
+            else:
+                tx_count += 1
+
+            tx_meta = tx.get_metadata()
+
+            # feed each transaction to the indexes that they are interested in
+            for index_filter, index in indexes_to_init:
```

Comment:
I feel we should call `init_loop_step(...)` for all transactions, and let the index decide whether it should skip it or not.

Reply:
This goes back to the other comment above: https://github.com/HathorNetwork/hathor-core/pull/392/files#r878182397

Although I think that would be better in the end, for now I think this is better. The implementation is also self-contained to `hathor/indexes/manager.py` and no details leak. The way it is right now makes it so the only place that has a decision on what type of transaction goes into an index is private and within `hathor/indexes/manager.py`. I prefer to avoid simultaneously changing both: how indexes are implemented and how indexes are managed. This PR has to change how indexes are managed, and I think it's best to do that while _not_ changing how indexes are implemented.

But I think it would be great if a future PR refactored this and move that to inside the index.

Reply:
Got it. I agree we can do it in a separate PR. Thanks!